<a href="https://colab.research.google.com/github/NatyDV/Business-Analytics/blob/main/Reto1_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1- Car Crash - Naive Bayes**
0. Cargar las librerias de trabajo

In [ ]:
import numpy as np #Funciones matemáticas
import pandas as pd #Conexión con archivos de Excel

from sklearn.naive_bayes import GaussianNB  #Modelo Gaussian Naive Bayes
from sklearn.metrics import confusion_matrix #Metricas- Como lo voy a evaluar

1. Se procede con la carga de los datos

In [ ]:
nxl="/content/1. BD2_CarCrash.xlsx"
XDB=pd.read_excel(nxl,sheet_name=3)

XD=XDB[['Weekday','ViolCat','ClearWeather','Month','Highway','Daylight' ]]
XDB=XDB.dropna() #Para evitar celdas vacias
XDB.head(100)

#Creamos el DataMart
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','Daylight']] #Variables Entrada
yd=XDB[['Highway']] #Valor Deseado- Aprenda el modelo


2. Se procede con la configuración y evaluación del modelo Naive Bayes (Car Crash )

In [ ]:
mnb=GaussianNB()
mnb.fit(XD,yd)

#Analizamos los resulatdos arrojados por el modelo
u=mnb.theta_ #Son las medias
print("Las medias de las variables son:\n ",u)
sigma=np.sqrt(mnb.var_) #Obtengo las desviaciones
print("Los valores de las desviaciones son:\n",sigma)

#Se procede con el análisis de las variaciones de las variables
LS=u+sigma
print("Los limites superiores para las variables son:\n ",LS)
LI=u-sigma
print("Los limites inferiores para las variables son:\n ",LI)

#Prediciendo los valores de preaprobación del valor deseado
ydp=mnb.predict(XD) #Hizo el pronóstico de los accidentes provocados en autopista

#se procede con la evaluación del comportamiento del modelo
cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es \n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VP+VN)/len(XD) #Exactitud
print("La Exactitud del Modelo es:",Ex)

Ter=(FP+FN)/len(XD) #Tasa de error
print("La Tasa de Error del Modelo es:",Ter)

Sen=VP/(VP+FN) #Sensibilidad
print("La Sensibilidad del modelos es:",Sen)

Esp=VN/(VN+FP) #Especificidad
print("La Especificidad del modelo es:",Esp)

Pres=VP/(VP+FP) #Precisión
print("La Precisión del modelo es:",Pres)

PredNeg=VN/(VN+FN) #Predicción del negativo
print("La Predicción negativo del modelo es :",PredNeg)

# Número total de datos
total_data = len(XDB)
print("Número total de datos es:",total_data)

np.sum(yd)/len(yd)  #Accidentes en Autopista
1-np.sum(yd)/len(yd) #Accidentes no en Autopista

#Para evaluar si el accidente fue en la autopista
X=np.array([[7,1,1,3,0]])
ydsc=mnb.predict(X)
print("El lugar del accidente es :",ydsc)


Las medias de las variables son:
  [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
Los valores de las desviaciones son:
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
Los limites superiores para las variables son:
  [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
Los limites inferiores para las variables son:
  [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es 
 [[4560  325]
 [1563  257]]
La Exactitud del Modelo es: 0.7184190902311708
La Tasa de Error del Modelo es: 0.2815809097688292
La Sensibilidad del modelos es: 0.1412087912087912
La Especificidad del modelo es: 0.9334698055271239
La Precisión del modelo es: 0.4415807560137457
La Predicción negativo del modelo es : 0.74473297

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return reduction(axis=axis, out=out, **passkwargs)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**Análisis de Resultado** Los resultados arrojados por el modelo propuesto frente al pronóstico de los accidentes en la autopista mostro  una precisión del 44%, lo que indica que 44% de los accidentes positivos han sidos predichos como positivos.La base de datos tiene un total de 6705 registros, de los cuales el 27.14% corresponden a accidentes en autopistas y el 72.86% a accidentes en otras vías.Adicionalmente La variable "Weekday" tiene un límite inferior alrededor de 2 y un límite superior alrededor de 5.8, lo que representa que los días más representativos para los accidentes están principalmente en la mitad de la semana. La variable "ViolCat" tiene un límite inferior de  3.14 y un límite superior de  10.41, lo que significa que la mayoría de los accidentes varían en severidad desde moderada hasta muy severa, con pocos accidentes muy leves (debajo de 3.14) y muy pocos extremadamente severos (por encima de 10.41). La exactitud del modelo es 71%, esto implica que el modelo tiene una veracidad de los datos razonablemente alta para la clasificación general, aunque su precisión específica para los accidentes en autopistas es menor.  La tasa de error es 28%, lo que indica que el modelo se equivoca en 28 de cada 100 predicciones realizadas. La sensibilidad del modelo es 14%, lo que refleja su baja capacidad para identificar correctamente los accidentes que ocurrieron en autopistas. La especificidad es 93%, indicando que el modelo es muy eficaz para identificar los accidentes que no ocurrieron en autopistas. Finalmente, para una persona con las características ['Weekday', 'ViolCat', 'ClearWeather', 'Month', 'Daylight']=[7,1,1,3,0], el modelo predice que vive en áreas rurales.